packages/core/src/extensions/Blocks/api/blockTypes.ts

```diff
@@ -0,0 +1,81 @@
+import { StyledText } from "./styleTypes";
+
+export type BlockTemplate<
+  // Type of the block.
+  // Examples might include: "paragraph", "heading", or "bulletListItem".
+  Type extends string,
+  // Changeable props which affect the block's behaviour or appearance.
+  // An example might be: { textAlignment: "left" | "right" | "center" } for a paragraph block.
+  Props extends Record<string, string>
+> = {
+  id: string;
+  type: Type;
+  props: Props;
+  content: StyledText[];
+  children: Block[];
+};
+
+export type PartialBlockTemplate<Template> = Template extends BlockTemplate<
+  infer Type,
+  infer Props
+>
+  ? {
+      id?: string;
+      type: Type;
+      props?: Partial<Props>;
+      content?: string | StyledText[];
+      children?: PartialBlock[];
+    }
+  : never;
+
+export type BlockPropsTemplate<Props> = Props extends Block["props"]
+  ? keyof Props
+  : never;
+
+export type GlobalProps = {
+  backgroundColor: string;
+  textColor: string;
+  textAlignment: "left" | "center" | "right" | "justify";
+};
+
+export type NumberedListItemBlock = BlockTemplate<
+  "numberedListItem",
+  GlobalProps
+>;
+
+export type BulletListItemBlock = BlockTemplate<"bulletListItem", GlobalProps>;
+
+export type HeadingBlock = BlockTemplate<
+  "heading",
+  GlobalProps & {
+    level: "1" | "2" | "3";
+  }
+>;
+
+export type ParagraphBlock = BlockTemplate<"paragraph", GlobalProps>;
+
+export type Block =
+  | ParagraphBlock
+  | HeadingBlock
+  | BulletListItemBlock
+  | NumberedListItemBlock;
+
+// @ts-ignore
```

Comment:
why ts-ignore?

Reply:
It's because otherwise TypeScript doesn't know for sure that `PartialBlock` doesn't recursively define itself. AFAIK it's ok when defining an object type to have fields in it of the same type, which is why it's fine that `Block`s contain more `Block`s inside them. For some reason though, `PartialBlock` being conditional breaks this.

The reason this didn't come up before is because `PartialBlock["children"]` was of type `Block[]`, which was incorrect.